File: solution/java/day15/src/main/java/document/DocumentTemplateType.java
package document;

import io.vavr.collection.List;
import io.vavr.collection.Map;

public enum DocumentTemplateType {
    DEERPP("DEER", RecordType.INDIVIDUAL_PROSPECT),
    DEERPM("DEER", RecordType.LEGAL_PROSPECT),
    AUTP("AUTP", RecordType.INDIVIDUAL_PROSPECT),
    AUTM("AUTM", RecordType.LEGAL_PROSPECT),
    SPEC("SPEC", RecordType.ALL),
    GLPP("GLPP", RecordType.INDIVIDUAL_PROSPECT),
    GLPM("GLPM", RecordType.LEGAL_PROSPECT);

    private static final Map<String, DocumentTemplateType> mapping =
            List.of(DocumentTemplateType.values())
                    .toMap(v -> formatKey(v.getDocumentType(), v.getRecordType().name()), v -> v)
                    .merge(List.of(RecordType.values()).toMap(v -> formatKey(SPEC.name(), v.name()), v -> SPEC));

    private final String documentType;
    private final RecordType recordType;

    DocumentTemplateType(String documentType, RecordType recordType) {
        this.documentType = documentType;
        this.recordType = recordType;
    }

    private static String formatKey(String documentType, String recordType) {
        return documentType.toUpperCase() + "-" + recordType.toUpperCase();
    }

    public static DocumentTemplateType fromDocumentTypeAndRecordType(String documentType, String recordType) {
        return mapping.get(formatKey(documentType, recordType))
                .getOrElseThrow(() -> new IllegalArgumentException("Invalid Document template type or record type"));
    }

    private RecordType getRecordType() {
        return recordType;
    }

    private String getDocumentType() {
        return documentType;
    }
}

File: solution/kotlin/day15/src/main/kotlin/document/DocumentTemplateType.kt
package document

import document.RecordType.*

enum class DocumentTemplateType(val documentType: String, val recordType: RecordType) {
    DEERPP("DEER", INDIVIDUAL_PROSPECT),
    DEERPM("DEER", LEGAL_PROSPECT),
    AUTP("AUTP", INDIVIDUAL_PROSPECT),
    AUTM("AUTM", LEGAL_PROSPECT),
    SPEC("SPEC", ALL),
    GLPP("GLPP", INDIVIDUAL_PROSPECT),
    GLPM("GLPM", LEGAL_PROSPECT);

    companion object {
        private val mapping: Map<String, DocumentTemplateType> = entries
            .associateBy { entry -> formatKey(entry.documentType, entry.recordType.name) }
            .plus(RecordType.entries.map { entry -> formatKey(SPEC.name, entry.name) to SPEC })

        private fun formatKey(documentType: String, recordType: String) =
            "${documentType.uppercase()}-${recordType.uppercase()}"

        fun fromDocumentTypeAndRecordType(documentType: String, recordType: String): DocumentTemplateType =
            mapping.getOrElse(
                formatKey(documentType, recordType)
            ) { throw IllegalArgumentException("Invalid Document template type or record type") }
    }
}